reana_client/api/client.py

```diff
@@ -420,3 +430,95 @@ def get_analysis_code(self, user, organization, analysis_id):
             raise Exception(e.response.json()['message'])
         except Exception as e:
             raise e
+
+    def upload_to_server(self, user, organization, workflow,
+                         filenames, upload_type):
+        """Upload file or directory to REANA-Server.
```

Comment:
Regarding the work on supporting directory upload,  perhaps we should be using "pathname" instead of "filename" in arguments etc, in order to better indicate that the inputs could be files or directiories?

We should beware of situations where the same file name can be nested in several directories:

```
code/filtering/run.py
code/generating/run.py
code/something/even/deeper/nested/run.py
```
We should properly define the behaviour, just reading the file from that local place, or even creating the subdirectory structure on the server[1]. This is basically why we chose with @hjhsalo to simply say "directory upload not supported yet" when the input argument were a directory.

[1] A technique I was thinking about regarding subdirectory upload could be something simple like `rsync`-ing files with their full parent directories to the destination place, e.g. by zero-zipping the input directory on the client and zero-unzipping it on the server.  Could be useful in case there are tons of inputs files. Could speak IRL about this.)

Reply:
Quick summary of IRL discussion:

- two nested file upload cases:

  ``r-c type upload a/b/c/foo.ext -> type/a/b/c/foo.ext OR type/b/c/foo.ext if type==a``

- two directory walk through upload cases:

  ``r-c type upload dirname/... -> type/dirname/... OR type/... if type==dirname``

Beware of mkdir-p-style creation of nested directiories in r-w-controller and beware of `../../..` like arguments. 

This should accommodate people using the recommended structure:

```
./code/mycode.py
./inputs/mydata.csv
```

or any other structure such as flat one:

```
./mycode.py
./mydata.csv
```

or arbitrary nested one:


```
./measurements/paris-2017/test-run-12b/mydata.csv
./thesis/chapter7/scripts/mycode.py
```

In the latter cases, we would recommend using `inputs` if people don't need the runtime code distinction, but that's another story where `type` could be eventually even left out...

Reply:
I think the `OR type/... if type==dirname` parts are not implemented yet.  

I tried running the ROOT6 example and I saw:

```
/reana/default/00000000-0000-0000-0000-000000000000/analyses/f1f421fc-b6cb-4c3f-9ec7-47c02e55ec71/yadage_inputs
/reana/default/00000000-0000-0000-0000-000000000000/analyses/f1f421fc-b6cb-4c3f-9ec7-47c02e55ec71/yadage_inputs/code
/reana/default/00000000-0000-0000-0000-000000000000/analyses/f1f421fc-b6cb-4c3f-9ec7-47c02e55ec71/yadage_inputs/code/code
/reana/default/00000000-0000-0000-0000-000000000000/analyses/f1f421fc-b6cb-4c3f-9ec7-47c02e55ec71/yadage_inputs/code/code/fitdata.C
/reana/default/00000000-0000-0000-0000-000000000000/analyses/f1f421fc-b6cb-4c3f-9ec7-47c02e55ec71/yadage_inputs/code/code/gendata.C
/reana/default/00000000-0000-0000-0000-000000000000/analyses/f1f421fc-b6cb-4c3f-9ec7-47c02e55ec71/yadage_inputs/inputs
/reana/default/00000000-0000-0000-0000-000000000000/analyses/f1f421fc-b6cb-4c3f-9ec7-47c02e55ec71/code
/reana/default/00000000-0000-0000-0000-000000000000/analyses/f1f421fc-b6cb-4c3f-9ec7-47c02e55ec71/code/code
/reana/default/00000000-0000-0000-0000-000000000000/analyses/f1f421fc-b6cb-4c3f-9ec7-47c02e55ec71/code/code/fitdata.C
/reana/default/00000000-0000-0000-0000-000000000000/analyses/f1f421fc-b6cb-4c3f-9ec7-47c02e55ec71/code/code/gendata.C
```

Note the `./code/code/...` nested directory.  This makes the example not run.

We should upload `./code/...` in this case since type==code.

Reply:
I think this is in: https://github.com/reanahub/reana-client/pull/79/files#diff-4f27f08c3f46f0b755c1d5ac1a1e2dcaR501
The check if the first directory is equal to the upload type is done only for files, as if it is a directory it will walk it, and call the function again, eventually with the file path.

When uploading with the command from the example, it returns an error because it doesn't assume I'm pointing to the code/ dir:
```
reana-client code upload fitdata.C
[Errno 2] No such file or directory: 'fitdata.C'
```

By specifying the code/ dir:
```
reana-client code upload code/fitdata.C
code/fitdata.C was uploaded successfully.

reana-client code upload code/gendata.C 
code/gendata.C was uploaded successfully.
```

and in the `workflow-controller` container the files are at the path:
```
/reana/default/00000000-0000-0000-0000-000000000000/analyses/b6cce0ad-569c-4cdf-a17e-fee6204ec5e1/code
```

If uploading the whole code/ dir with:
```
reana-client code upload code
code/gendata.C was uploaded successfully.
code/fitdata.C was uploaded successfully.
```
 they are again at the same path.

If we upload sth outside of the code/ directory:
```
reana-client code upload a.txt
a.txt was uploaded successfully.
```
It'll be placed in code/ in the `workflow-controller` and the example runs ok.
Maybe I'm missing sth else, but I think this should be ok.